File: py/slop/extract.py
"""Turn a source file into prose plus a map back to source offsets.

Every extractor returns (text, runs). `text` is prose with markup removed;
`to_source(runs, offset)` converts an offset in it back to a byte offset in the
original file, so findings report accurate file:line:col.

Three rules drive the design, all of them learned from tests/run.mjs:

1. Inline markup is deleted, not blanked, so "in<b>ter</b>nal" stays one word.
2. Block boundaries emit a newline, but inline whitespace does not. HTML
   collapses whitespace, and markdown soft-wraps, so a newline in the *source*
   inside a paragraph is a space -- treating it as a line break puts a false
   sentence boundary at every wrap column.
3. Long unbroken lines are hard-wrapped. The sentence detectors rescan from
   every start position, so text with no terminator costs O(n^2).
"""
from __future__ import annotations

import html as _html
import re
from html.parser import HTMLParser

WRAP = 2000


class Mapper:
    """Builds extracted text while remembering where each piece came from."""

    def __init__(self, wrap: int = WRAP):
        self._chunks: list[str] = []
        self.runs: list[tuple[int, int, int, bool]] = []  # (out, src, len, fixed)
        self.len = 0
        self.col = 0
        self.wrap = max(80, wrap)

    def _raw(self, text: str, src: int, fixed: bool = False) -> None:
        if not text:
            return
        self._chunks.append(text)
        self.runs.append((self.len, src, len(text), fixed))
        self.len += len(text)

    def copy(self, text: str, src_start: int) -> None:
        """Text copied verbatim from the source; offsets map 1:1."""
        if not text:
            return
        i = 0
        while i < len(text):
            nl = text.find("\n", i)
            seg_end = len(text) if nl == -1 else nl + 1
            j = i
            while j < seg_end:
                room = max(1, self.wrap - self.col)
                if seg_end - j <= room:
                    self._raw(text[j:seg_end], src_start + j)
                    self.col += seg_end - j
                    j = seg_end
                else:
                    limit = min(j + room, seg_end)
                    br = -1
                    for k in range(limit - 1, j, -1):
                        if text[k] in " \t":
                            br = k
                            break
                    if br == -1:
                        br = limit
                    self._raw(text[j:br], src_start + j)
                    self._raw("\n", src_start + br, True)
                    self.col = 0
                    j = br
            if nl != -1:
                self.col = 0
            i = seg_end

    def sub(self, text: str, src_pos: int) -> None:
        """Text substituted for the source; every char maps to one offset."""
        if not text:
            return
        self._raw(text, src_pos, True)
        nl = text.rfind("\n")
        self.col = self.col + len(text) if nl == -1 else len(text) - nl - 1

    def brk(self, src_pos: int) -> None:
        """Emit a block separator, collapsing runs of them."""
        if self.len == 0 or (self._chunks and self._chunks[-1].endswith("\n")):
            return
        self.sub("\n", src_pos)

    def build(self) -> tuple[str, list]:
        return "".join(self._chunks), self.runs


def to_source(runs, off: int) -> int:
    if not runs:
        return 0
    lo, hi = 0, len(runs) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        o, s, n, fixed = runs[mid]
        if off < o:
            hi = mid - 1
        elif off >= o + n:
            lo = mid + 1
        else:
            return s if fixed else s + (off - o)
    o, s, n, fixed = runs[-1]
    return s + (0 if fixed else n)


# ------------------------------------------------------------------- HTML

SKIP_EL = {"script", "style", "pre", "code", "kbd", "samp", "var", "tt", "svg",
           "math", "textarea", "template", "noscript", "iframe", "object",
           "canvas", "head"}
BLOCK_EL = {"address", "article", "aside", "blockquote", "br", "caption", "dd",
            "details", "dialog", "div", "dl", "dt", "fieldset", "figcaption",
            "figure", "footer", "form", "h1", "h2", "h3", "h4", "h5", "h6",
            "header", "hgroup", "hr", "legend", "li", "main", "menu", "nav",
            "ol", "p", "section", "summary", "table", "tbody", "td", "tfoot",
            "th", "thead", "tr", "ul", "option", "title", "body", "html"}

_WS_RUN = re.compile(r"\s*\n\s*|[ \t]{2,}")


def _copy_html_text(mapper: Mapper, text: str, base: int) -> None:
    last = 0
    for m in _WS_RUN.finditer(text):
        if m.start() > last:
            mapper.copy(text[last:m.start()], base + last)
        mapper.sub(" ", base + m.start())
        last = m.end()
    if last < len(text):
        mapper.copy(text[last:], base + last)


class _HtmlExtractor(HTMLParser):
    def __init__(self, src: str, mapper: Mapper):
        super().__init__(convert_charrefs=False)
        self.src = src
        self.mapper = mapper
        self.skip = 0
        self._line_starts = [0]
        for i, ch in enumerate(src):
            if ch == "\n":
                self._line_starts.append(i + 1)

    def _off(self) -> int:
        line, col = self.getpos()
        return self._line_starts[line - 1] + col

    def handle_starttag(self, tag, attrs):
        if tag in SKIP_EL:
            self.skip += 1
            self.mapper.brk(self._off())
        elif tag in BLOCK_EL:
            self.mapper.brk(self._off())

    def handle_startendtag(self, tag, attrs):
        if tag in BLOCK_EL:
            self.mapper.brk(self._off())

    def handle_endtag(self, tag):
        if tag in SKIP_EL:
            self.skip = max(0, self.skip - 1)
            self.mapper.brk(self._off())
        elif tag in BLOCK_EL:
            self.mapper.brk(self._off())

    def handle_data(self, data):
        if self.skip:
            return
        _copy_html_text(self.mapper, data, self._off())

    def handle_entityref(self, name):
        if self.skip:
            return
        self.mapper.sub(_html.unescape("&" + name + ";"), self._off())

    def handle_charref(self, name):
        if self.skip:
            return
        self.mapper.sub(_html.unescape("&#" + name + ";"), self._off())

    # A comment is inline: "<p>a <!-- x --> b</p>" is one sentence.
    def handle_comment(self, data):
        pass

    def handle_decl(self, decl):
        self.mapper.brk(self._off())

    def handle_pi(self, data):
        self.mapper.brk(self._off())


def extract_html(src: str, **opts) -> tuple[str, list]:
    mapper = Mapper(opts.get("wrap") or WRAP)
    parser = _HtmlExtractor(src, mapper)
    parser.feed(src)
    parser.close()
    return mapper.build()


# --------------------------------------------------------------- Markdown

FENCE = re.compile(r"^(\s{0,3})(`{3,}|~{3,})(.*)$")
SETEXT = re.compile(r"^\s{0,3}(?:=+|-{2,})\s*$")
ATX = re.compile(r"^(\s{0,3}#{1,6}\s+)")
QUOTE = re.compile(r"^(\s*(?:>\s?)+)")
BULLET = re.compile(r"^(\s*(?:[-*+]|\d{1,9}[.)])\s+)")
REFDEF = re.compile(r"^\s{0,3}\[[^\]]+\]:\s*\S")
HRULE = re.compile(r"^\s{0,3}(?:(?:\*\s*){3,}|(?:-\s*){3,}|(?:_\s*){3,})$")
TABLE_DELIM = re.compile(r"^\s*\|?[\s:|-]*-[\s:|-]*\|[\s:|-]*$")
TABLE_ROW = re.compile(r"^\s*\|.*\|\s*$")

MD_INLINE = re.compile(
    r"(`+)[\s\S]*?\1"                                             # code span
    r"|!?\[((?:[^\[\]]|\[[^\]]*\])*)\]\((?:[^()\\]|\\.|\([^)]*\))*\)"  # inline link
    r"|!?\[((?:[^\[\]]|\[[^\]]*\])*)\]\[[^\]]*\]"                 # reference link
    r"|\[\^[^\]]+\]"                                              # footnote ref
    r"|<(?:https?|mailto|ftp):[^>\s]*>"                           # autolink
    r"|</?[a-zA-Z][^>]*>"                                         # raw HTML tag
    r"|(?:https?://|www\.)[^\s<>()\[\]\"']+"                      # bare URL
    r"|\*{1,3}|~~|={2,}"                                          # emphasis runs
    r"|(?<![A-Za-z0-9])_{1,3}|_{1,3}(?![A-Za-z0-9])"              # _emphasis_, not snake_case
)


def _md_inline(mapper: Mapper, line: str, base: int, depth: int = 0) -> None:
    last = 0
    pos = 0
    while pos <= len(line):
        m = MD_INLINE.search(line, pos)
        if not m:
            break
        if m.start() < last:
            pos = last
            continue
        text = m.group(2) if m.group(2) is not None else m.group(3)
        gap_end, after = m.start(), m.end()
        # A dropped token inside brackets would leave "()" behind.
        if not text:
            if (gap_end and after < len(line)
                    and ((line[gap_end - 1] == "(" and line[after] == ")")
                         or (line[gap_end - 1] == "[" and line[after] == "]"))):
                gap_end -= 1
                after += 1
        if gap_end > last:
            mapper.copy(line[last:gap_end], base + last)
        if text:
            at = m.group(0).find("[" + text)
            if at != -1:
                off = m.start() + at + 1
                if depth < 3:
                    _md_inline(mapper, text, base + off, depth + 1)
                else:
                    mapper.copy(text, base + off)
        last = after
        pos = max(after, m.start() + 1)
    if last < len(line):
        mapper.copy(line[last:], base + last)


def extract_markdown(src: str, **opts) -> tuple[str, list]:
    mapper = Mapper(opts.get("wrap") or WRAP)
    indent_code = opts.get("indent_code", True)
    skip_tables = opts.get("skip_tables", False)

    lines, pos = [], 0
    for raw in src.split("\n"):
        lines.append((raw.rstrip("\r"), pos))
        pos += len(raw) + 1

    i = 0
    if lines and re.match(r"^(---|\+\+\+)\s*$", lines[0][0]):
        close = lines[0][0].strip()
        for k in range(1, len(lines)):
            if lines[k][0].strip() == close:
                i = k + 1
                break

    fence = None
    recent_list = 0
    is_open = False
    prev_ends_block = True

    def close_block(off: int) -> None:
        nonlocal is_open
        if is_open:
            mapper.sub("\n", off)
            is_open = False

    while i < len(lines):
        line, off = lines[i]
        i += 1

        if fence is not None:
            f = FENCE.match(line)
            if f and f.group(2)[0] == fence[0] and len(f.group(2)) >= len(fence) and not f.group(3).strip():
                fence = None
            close_block(off)
            continue
        f = FENCE.match(line)
        if f:
            fence = f.group(2)
            close_block(off)
            continue
        if not line.strip():
            close_block(off)
            continue
        if indent_code and re.match(r"^(?: {4,}|\t)", line) and recent_list == 0:
            close_block(off)
            continue
        if REFDEF.match(line) or HRULE.match(line) or SETEXT.match(line) or TABLE_DELIM.match(line):
            close_block(off)
            continue
        if skip_tables and TABLE_ROW.match(line):
            close_block(off)
            continue

        recent_list = 3 if BULLET.match(line) else max(0, recent_list - 1)

        start = 0
        for rx in (QUOTE, ATX, BULLET):
            m = rx.match(line[start:])
            if m:
                start += len(m.group(1))
        q = QUOTE.match(line)
        after_quote = line[len(q.group(1)):] if q else line
        is_heading = bool(ATX.match(after_quote))
        is_table = bool(TABLE_ROW.match(line))
        starts_block = is_heading or is_table or bool(BULLET.match(after_quote))

        if is_open:
            mapper.sub("\n" if (starts_block or prev_ends_block) else " ", off)

        body = line[start:]
        if is_table:
            body = body.replace("|", " ")
        _md_inline(mapper, body, off + start)
        is_open = True
        prev_ends_block = is_heading or is_table or bool(re.search(r"(?: {2,}|\\)$", line))

    close_block(len(src))
    return mapper.build()


def extract_plain(src: str, **opts) -> tuple[str, list]:
    mapper = Mapper(opts.get("wrap") or WRAP)
    mapper.copy(src, 0)
    return mapper.build()


EXTRACTORS = {
    "html": extract_html, "htm": extract_html, "xhtml": extract_html,
    "md": extract_markdown, "markdown": extract_markdown,
    "mdx": extract_markdown, "mdown": extract_markdown,
}


def extractor_for(name: str):
    ext = name.rsplit(".", 1)[-1].lower() if "." in name else ""
    return EXTRACTORS.get(ext, extract_plain)

File: js/extract.mjs
// Text extraction for cliche-lint.
//
// Both extractors return { text, runs }: `text` is prose with markup removed,
// `runs` maps every offset in it back to a byte offset in the original file so
// findings can be reported as file:line:col.
//
// Two rules drive the design:
//  1. Inline markup is *deleted*, not blanked, so "in<b>ter</b>nal" stays one
//     word instead of becoming three.
//  2. Block boundaries *must* emit a newline. The upstream sentence regexes
//     (/[^.!?\n]+[.!?]/ and friends) go quadratic on a long run of text with no
//     terminator: 8000 words on one line takes ~4.3s vs ~30ms when broken into
//     lines. Flattening HTML without newlines is the easy way to hit that.

const WRAP = 2000; // safety net: hard-wrap any line longer than this

export class Mapper {
  constructor(wrap = WRAP) {
    this.chunks = [];
    this.runs = [];
    this.len = 0;
    this.col = 0;
    this.wrap = Math.max(80, wrap);
  }

  _raw(text, srcStart, fixed) {
    if (!text) return;
    this.chunks.push(text);
    this.runs.push({ o: this.len, s: srcStart, n: text.length, fixed: !!fixed });
    this.len += text.length;
  }

  // Text copied verbatim from the source; offsets map 1:1.
  copy(text, srcStart) {
    if (!text) return;
    let i = 0;
    while (i < text.length) {
      const nl = text.indexOf('\n', i);
      const segEnd = nl === -1 ? text.length : nl + 1;
      let j = i;
      while (j < segEnd) {
        const room = Math.max(1, this.wrap - this.col);
        if (segEnd - j <= room) {
          this._raw(text.slice(j, segEnd), srcStart + j);
          this.col += segEnd - j;
          j = segEnd;
        } else {
          const limit = Math.min(j + room, segEnd);
          let br = -1;
          for (let k = limit - 1; k > j; k--) {
            if (text[k] === ' ' || text[k] === '\t') { br = k; break; }
          }
          if (br === -1) br = limit;
          this._raw(text.slice(j, br), srcStart + j);
          this._raw('\n', srcStart + br, true);
          this.col = 0;
          j = br;
        }
      }
      if (nl !== -1) this.col = 0;
      i = segEnd;
    }
  }

  // Text substituted for the source (decoded entity, synthetic newline);
  // every character maps to the same source offset.
  sub(text, srcPos) {
    if (!text) return;
    this._raw(text, srcPos, true);
    const nl = text.lastIndexOf('\n');
    if (nl === -1) this.col += text.length;
    else this.col = text.length - nl - 1;
  }

  // Emit a block separator, collapsing runs of them.
  brk(srcPos) {
    const t = this.chunks[this.chunks.length - 1];
    if (this.len === 0 || (t && t.endsWith('\n'))) return;
    this.sub('\n', srcPos);
  }

  build() {
    return { text: this.chunks.join(''), runs: this.runs };
  }
}

// Map an offset in the extracted text back to a source offset.
export function toSource(runs, off) {
  if (!runs.length) return 0;
  let lo = 0;
  let hi = runs.length - 1;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    const r = runs[mid];
    if (off < r.o) hi = mid - 1;
    else if (off >= r.o + r.n) lo = mid + 1;
    else return r.fixed ? r.s : r.s + (off - r.o);
  }
  const last = runs[runs.length - 1];
  return last.s + (last.fixed ? 0 : last.n);
}

// ---------------------------------------------------------------- entities

const NAMED = {
  amp: '&', lt: '<', gt: '>', quot: '"', apos: "'", nbsp: ' ',
  mdash: '—', ndash: '–', hellip: '…', rsquo: '’',
  lsquo: '‘', ldquo: '“', rdquo: '”', copy: '©',
  reg: '®', trade: '™', deg: '°', hyphen: '‐',
  laquo: '«', raquo: '»', middot: '·', bull: '•',
  times: '×', minus: '−', eacute: 'é', egrave: 'è',
  ouml: 'ö', uuml: 'ü', auml: 'ä', szlig: 'ß',
  euro: '€', pound: '£', shy: '', zwnj: '', zwj: '', ensp: ' ',
  emsp: ' ', thinsp: ' ', lsaquo: '‹', rsaquo: '›', dagger: '†'
};
const ENTITY = /&(#[0-9]+|#[xX][0-9a-fA-F]+|[a-zA-Z][a-zA-Z0-9]{1,31});/g;

function decodeInto(mapper, text, base) {
  ENTITY.lastIndex = 0;
  let last = 0;
  let m;
  while ((m = ENTITY.exec(text))) {
    if (m.index > last) mapper.copy(text.slice(last, m.index), base + last);
    const body = m[1];
    let out;
    if (body[0] === '#') {
      const code = body[1] === 'x' || body[1] === 'X'
        ? parseInt(body.slice(2), 16)
        : parseInt(body.slice(1), 10);
      out = Number.isFinite(code) && code > 0 && code <= 0x10ffff ? String.fromCodePoint(code) : m[0];
    } else {
      out = Object.prototype.hasOwnProperty.call(NAMED, body) ? NAMED[body] : m[0];
    }
    mapper.sub(out, base + m.index);
    last = m.index + m[0].length;
  }
  if (last < text.length) mapper.copy(text.slice(last), base + last);
}

// HTML collapses whitespace runs in normal flow, so a newline inside a <p> is
// a space, not a sentence boundary. Collapsing before matching keeps the
// sentence-based detectors from seeing a false break at the source wrap point.
const WS_RUN = /\s*\n\s*|[ \t]{2,}/g;

function copyHtmlText(mapper, text, base) {
  WS_RUN.lastIndex = 0;
  let last = 0;
  let m;
  while ((m = WS_RUN.exec(text))) {
    if (m.index > last) decodeInto(mapper, text.slice(last, m.index), base + last);
    mapper.sub(' ', base + m.index);
    last = m.index + m[0].length;
  }
  if (last < text.length) decodeInto(mapper, text.slice(last), base + last);
}

// -------------------------------------------------------------------- HTML

const SKIP_EL = /^(?:script|style|pre|code|kbd|samp|var|tt|svg|math|textarea|template|noscript|iframe|object|canvas|head)$/i;
const BLOCK_EL = /^(?:address|article|aside|blockquote|br|caption|dd|details|dialog|div|dl|dt|fieldset|figcaption|figure|footer|form|h[1-6]|header|hgroup|hr|legend|li|main|menu|nav|ol|p|section|summary|table|tbody|td|tfoot|th|thead|tr|ul|option|title|body|html)$/i;
const HTML_TOKEN = /<!--[\s\S]*?-->|<!\[CDATA\[[\s\S]*?\]\]>|<![^>]*>|<\?[\s\S]*?\?>|<\/?([a-zA-Z][a-zA-Z0-9:-]*)((?:"[^"]*"|'[^']*'|[^"'>])*)>/g;

export function extractHtml(src, opts = {}) {
  const mapper = new Mapper(opts.wrap);
  HTML_TOKEN.lastIndex = 0;
  let last = 0;
  let m;
  while ((m = HTML_TOKEN.exec(src))) {
    if (m.index > last) copyHtmlText(mapper, src.slice(last, m.index), last);
    last = m.index + m[0].length;
    const name = m[1];
    if (!name) {
      // A comment is inline: "<p>a <!-- x --> b</p>" is one sentence. Doctype
      // and processing instructions sit between blocks.
      if (!m[0].startsWith('<!--')) mapper.brk(m.index);
      continue;
    }
    const closing = m[0][1] === '/';
    const selfClosing = /\/\s*$/.test(m[2] || '');
    if (!closing && !selfClosing && SKIP_EL.test(name)) {
      // Jump past the matching close tag; these elements do not nest in practice.
      const close = new RegExp('</\\s*' + name.replace(/[^\w:-]/g, '') + '\\s*>', 'i');
      close.lastIndex = 0;
      const rest = src.slice(last);
      const hit = close.exec(rest);
      const end = hit ? last + hit.index + hit[0].length : src.length;
      mapper.brk(m.index);
      HTML_TOKEN.lastIndex = end;
      last = end;
      continue;
    }
    if (BLOCK_EL.test(name)) mapper.brk(m.index);
  }
  if (last < src.length) copyHtmlText(mapper, src.slice(last), last);
  return mapper.build();
}

// ---------------------------------------------------------------- Markdown

const FENCE = /^(\s{0,3})(`{3,}|~{3,})(.*)$/;
const SETEXT = /^\s{0,3}(?:=+|-{2,})\s*$/;
const ATX = /^(\s{0,3}#{1,6}\s+)/;
const QUOTE = /^(\s*(?:>\s?)+)/;
const BULLET = /^(\s*(?:[-*+]|\d{1,9}[.)])\s+)/;
const REFDEF = /^\s{0,3}\[[^\]]+\]:\s*\S/;
const HRULE = /^\s{0,3}(?:(?:\*\s*){3,}|(?:-\s*){3,}|(?:_\s*){3,})$/;
const TABLE_DELIM = /^\s*\|?[\s:|-]*-[\s:|-]*\|[\s:|-]*$/;
const TABLE_ROW = /^\s*\|.*\|\s*$/;

// Inline markup: code spans, links/images, raw HTML, autolinks, bare URLs,
// and emphasis runs. Link/image text is kept and re-scanned; everything else
// in the token is dropped.
const MD_INLINE = new RegExp([
  '(`+)[\\s\\S]*?\\1',                                        // code span
  '!?\\[((?:[^\\[\\]]|\\[[^\\]]*\\])*)\\]\\((?:[^()\\\\]|\\\\.|\\([^)]*\\))*\\)', // inline link
  '!?\\[((?:[^\\[\\]]|\\[[^\\]]*\\])*)\\]\\[[^\\]]*\\]',      // reference link
  '\\[\\^[^\\]]+\\]',                                         // footnote reference
  '<(?:https?|mailto|ftp):[^>\\s]*>',                          // autolink
  '</?[a-zA-Z][^>]*>',                                         // raw HTML tag
  '(?:https?://|www\\.)[^\\s<>()\\[\\]"\']+',                  // bare URL
  '\\*{1,3}|~~|={2,}',                                         // emphasis runs
  // Underscore emphasis, but never inside a word: _italic_ goes, snake_case stays.
  '(?<![A-Za-z0-9])_{1,3}|_{1,3}(?![A-Za-z0-9])'
].join('|'), 'g');

function mdInline(mapper, line, base, depth) {
  const re = new RegExp(MD_INLINE.source, 'g');
  let last = 0;
  let m;
  while ((m = re.exec(line))) {
    if (m.index < last) { re.lastIndex = last; continue; }
    const tok = m[0];
    const text = m[2] !== undefined ? m[2] : m[3];
    let gapEnd = m.index;
    let after = m.index + tok.length;
    // A dropped token inside brackets would leave "()" behind -- as in
    // "AX200 (`iwlwifi`)". Take the brackets with it.
    if (!text) {
      const open = line[gapEnd - 1];
      const close = line[after];
      if ((open === '(' && close === ')') || (open === '[' && close === ']')) { gapEnd -= 1; after += 1; }
    }
    if (gapEnd > last) mapper.copy(line.slice(last, gapEnd), base + last);
    if (text) {
      const at = tok.indexOf('[' + text);
      if (at !== -1) {
        const off = m.index + at + 1;
        if (depth < 3) mdInline(mapper, text, base + off, depth + 1);
        else mapper.copy(text, base + off);
      }
    }
    last = after;
    re.lastIndex = Math.max(after, m.index + 1);
  }
  if (last < line.length) mapper.copy(line.slice(last), base + last);
}

export function extractMarkdown(src, opts = {}) {
  const mapper = new Mapper(opts.wrap);
  const indentCode = opts.indentCode !== false;
  const skipTables = !!opts.skipTables;
  const lines = [];
  {
    let pos = 0;
    for (const raw of src.split('\n')) {
      lines.push({ text: raw.replace(/\r$/, ''), off: pos });
      pos += raw.length + 1;
    }
  }

  let i = 0;
  // YAML / TOML front matter
  if (lines.length && /^(---|\+\+\+)\s*$/.test(lines[0].text)) {
    const close = lines[0].text.trim();
    for (let k = 1; k < lines.length; k++) {
      if (lines[k].text.trim() === close) { i = k + 1; break; }
    }
  }

  let fence = null;
  let recentList = 0;
  let open = false;       // a prose line is emitted and awaiting its separator
  let prevEndsBlock = true;
  const close = (off) => { if (open) { mapper.sub('\n', off); open = false; } };

  for (; i < lines.length; i++) {
    const { text: line, off } = lines[i];

    if (fence) {
      const f = line.match(FENCE);
      if (f && f[2][0] === fence[0] && f[2].length >= fence.length && !f[3].trim()) fence = null;
      close(off);
      continue;
    }
    const f = line.match(FENCE);
    if (f) { fence = f[2]; close(off); continue; }

    if (!line.trim()) { close(off); continue; }

    // Indented code: 4+ spaces, not a list continuation.
    if (indentCode && /^(?: {4,}|\t)/.test(line) && recentList === 0) { close(off); continue; }

    if (REFDEF.test(line) || HRULE.test(line) || SETEXT.test(line) || TABLE_DELIM.test(line)) {
      close(off);
      continue;
    }
    if (skipTables && TABLE_ROW.test(line)) { close(off); continue; }

    recentList = BULLET.test(line) ? 3 : Math.max(0, recentList - 1);

    let start = 0;
    for (const re of [QUOTE, ATX, BULLET]) {
      const m = line.slice(start).match(re);
      if (m) start += m[1].length;
    }
    const afterQuote = line.slice((line.match(QUOTE) || [, ''])[1].length);
    const isHeading = ATX.test(afterQuote);
    const isTable = TABLE_ROW.test(line);
    const startsBlock = isHeading || isTable || BULLET.test(afterQuote);

    // Markdown soft-wraps: consecutive prose lines are one paragraph and join
    // with a space. Emitting a newline instead would put a false sentence
    // boundary at every hard-wrap column, which breaks the sentence detectors.
    if (open) mapper.sub(startsBlock || prevEndsBlock ? '\n' : ' ', off);

    let body = line.slice(start);
    // Table cells become spaces so a row reads as one line of prose.
    if (isTable) body = body.replace(/\|/g, ' ');
    mdInline(mapper, body, off + start, 0);
    open = true;
    prevEndsBlock = isHeading || isTable || /(?: {2,}|\\)$/.test(line);
  }
  close(src.length);
  return mapper.build();
}

export function extractPlain(src, opts = {}) {
  const mapper = new Mapper(opts.wrap);
  mapper.copy(src, 0);
  return mapper.build();
}

export function extractorFor(file) {
  const ext = (file.match(/\.([a-z0-9]+)$/i) || [, ''])[1].toLowerCase();
  if (ext === 'html' || ext === 'htm' || ext === 'xhtml') return extractHtml;
  if (ext === 'md' || ext === 'markdown' || ext === 'mdx' || ext === 'mdown') return extractMarkdown;
  return extractPlain;
}
